### PR TITLE
Restrict subnets only to subnets from regular availability zones in ELB auto-discovery procedure

### LINKED ISSUE #442: Introduce well-known tag for exclude subnets within a auto-discovery procedure for ELB backed services
**What would you like to be added**:

Another well known tag for subnets which will helps to exclude subnets and prevent its attachment to an ELB during auto-discovery procedure.

or 

extend `kubernetes.io/role/elb` semantic and allow to specify `kubernetes.io/role/elb=0` for subnets exclusion.

**Why is this needed**:

Currently subnets auto-discovery procedure for ELB relies on `kubernetes.io/cluster/{clusterId}` and/or `kubernetes.io/role/elb` tags, however it might be desirable to not attach subnets in certain zones (local zones, wavelength zones) but still keep `kubernetes.io/cluster/{clusterId}` for another automation purposes. 

Some context (Openshift specific unfortunately): https://bugzilla.redhat.com/show_bug.cgi?id=2105337

/kind feature

### PR DESCRIPTION
**Pull request by lobziik**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
Outpost, wavelength, and local zone don't support NLB or CLB at the moment. 
This pr adds check for availability zone type and includes only subnets from `availability-zone` typed zones for the auto-discovery.

**Which issue(s) this PR fixes**:

Fixes #442 

**Special notes for your reviewer**:
I'm not sure about PR type, guess that `feature` is the most suitable one.

**Does this PR introduce a user-facing change?**:

```release-note
Added check  for availability zone type within subnets auto-discovery procedure for a load-balancers
Auto-discovery will include only subnets from `availability-zone` typed zones.

IMPORTANT: This change requires an IAM permissions update. IAM policy for control-plane nodes
should be extended with `ec2:DescribeAvailabilityZones` permission.
```

### CODE REVIEW

**Comment by JoelSpeed** [comment]
/lgtm

Will need an AWS approver to review as well, but the code looks clean and makes sense to me

**Comment by lobziik** [comment]
e2e is failing, i need some time to reproduce and investigate it.
/hold

**Comment by lobziik** [comment]
Apparently this requires additional iam permissions for describing availability zones. I need to figure out where it does set within this e2es and give an another thought about this api call error handling.

Will appreciate some advices here, for now i'm not entirely sure how to proceed.

File: pkg/providers/v1/aws.go
Comment (mtulio) [inline pkg/providers/v1/aws.go:1169]: You may also need to update the [IAM Policy document](https://github.com/kubernetes/cloud-provider-aws/blob/master/docs/prerequisites.md#iam-policies) by adding the `ec2:DescribeAvailabilityZones` permission. That IAM Policy should be attached to the control plane IAM Instance Profile/Role.

Comment (lobziik) [reply pkg/providers/v1/aws.go:1169]: Updated

**Comment by lobziik** [comment]
Updated PR, added a check for the response type. Insufficient permissions does not threat as an error anymore, warning will be raised instead: https://github.com/kubernetes/cloud-provider-aws/pull/499/files#diff-97a31d0f193ea0b4e2c538595daf93405351054377458053a5010f6fdb97f4dcR3539 for provide backward compatability and do not break existing clusters after ccm upgrade.

Added missed permission into documentation also.

**Comment by lobziik** [comment]
/cc @kishorj @JoelSpeed 

Sry, I misclicked and break review requests.

File: pkg/providers/v1/aws.go
Comment (JoelSpeed) [inline pkg/providers/v1/aws.go:1170]: If you use `fmt.Errorf` with a `%w` format string then in later calls `errors.Is` can check this by unwrapping

File: pkg/providers/v1/aws.go
Comment (JoelSpeed) [inline pkg/providers/v1/aws.go:3539]: If you wrap the inner error then you could do something like this to achieve the same
```
awsErr := awserr.RequestFailure{}
if errors.As(err, awsErr) && awsErr.Code() ...
```

**Review by JoelSpeed — COMMENTED** [review]
Changes look ok, couple of nits about errors but otherwise LGTM, lmk what you think about using the error wrapping, this file already imports the standard library errors package so there's no new dependency on this

File: pkg/providers/v1/aws.go
Comment (lobziik) [reply pkg/providers/v1/aws.go:3539]: I looked over the file, and the only pattern to check aws errors is through type assertions. I would prefer to keep it consistent and not introduce a new pattern here. 

But i agree, wrap errors with "%w" and use `errors.As` will be better, but i suggest to address this in other PR and make this same across errors which is returning from this `ec2` sdk wrapper.

Comment (JoelSpeed) [reply pkg/providers/v1/aws.go:3539]: If this keeps it consistent then fair enough

Comment (kishorj) [inline pkg/providers/v1/aws.go:3539]: Tbh, if we ignore the error we will keep the existing behavior unless the IAM permissions get updated. I favor throwing the error here instead.

We will have to update the IAM permissions. As for EKS, we can start process to update the managed policies. For kops, we can request the permission to be included.

Comment (kishorj) [reply pkg/providers/v1/aws.go:3539]: Once the policies get updates, we can backport the changes.

**Comment by lobziik** [comment]
I attempted to update permission list for kops. @olemarkus can you please help with that, when you will have time? :) https://github.com/kubernetes/kops/pull/14650 - think it will be a prerequisite for having `pull-cloud-provider-aws-e2e` passed.

File: pkg/providers/v1/aws.go
Comment (lobziik) [reply pkg/providers/v1/aws.go]: I updated pr, now permissions error will be propagated back. 

IAM permissions in kops were extended in https://github.com/kubernetes/kops/pull/14650.

@kishorj can you please take a look? :)

**Comment by JoelSpeed** [comment]
Thanks for the fixups @lobziik 

/lgtm

**Comment by kishorj** [comment]
Thanks
/lgtm

Need updates to the AmazonEKSClusterPolicy for EKS clusters. Hold merge until we initiate the process to update the managed policy.

**Review by olemarkus — APPROVED** [review]

**Pull request merged**
